src/sources/aws_s3/sqs.rs

```diff
@@ -506,31 +514,65 @@ impl IngestorProcess {
             None => lines,
         };
 
-        let bucket_name = Bytes::from(s3_event.s3.bucket.name.as_str().as_bytes().to_vec());
-        let object_key = Bytes::from(s3_event.s3.object.key.as_str().as_bytes().to_vec());
-        let aws_region = Bytes::from(s3_event.aws_region.as_str().as_bytes().to_vec());
-
-        let mut stream = lines.filter_map(move |line| {
-            let mut log = LogEvent::from_bytes_legacy(&line).with_batch_notifier_option(&batch);
-
-            log.insert(event_path!("bucket"), bucket_name.clone());
-            log.insert(event_path!("object"), object_key.clone());
-            log.insert(event_path!("region"), aws_region.clone());
-            log.insert(log_schema().source_type_key(), Bytes::from("aws_s3"));
-            log.insert(log_schema().timestamp_key(), timestamp);
+        let mut stream = lines.map(|line| {
+            let deserializer = BytesDeserializer::new();
+            let mut log = deserializer
+                .parse_single(line, log_namespace)
+                .with_batch_notifier_option(&batch);
+
+            log_namespace.insert_source_metadata(
+                AwsS3Config::NAME,
+                &mut log,
+                path!("bucket"),
+                path!("bucket"),
+                Bytes::from(s3_event.s3.bucket.name.as_bytes().to_vec()),
+            );
+            log_namespace.insert_source_metadata(
+                AwsS3Config::NAME,
+                &mut log,
+                path!("object"),
+                path!("object"),
+                Bytes::from(s3_event.s3.object.key.as_bytes().to_vec()),
+            );
+            log_namespace.insert_source_metadata(
+                AwsS3Config::NAME,
+                &mut log,
+                path!("region"),
+                path!("region"),
+                Bytes::from(s3_event.aws_region.as_bytes().to_vec()),
+            );
 
             if let Some(metadata) = &metadata {
                 for (key, value) in metadata {
-                    log.insert(key.as_str(), value.clone());
+                    log_namespace.insert_source_metadata(
```

Comment:
When using the new `vector` namespace, we must _guarantee_ that fields cannot collide. Since the keys here are from S3 (Vector can't control them) they need to be nested under a known key (such as `metadata`).

For example, there is currently a `region` source metadata being inserted by Vector. If the S3 object _also_ had a `region` key, this would override the previous key. Not only does this delete the previous data so users can't use it, but it could cause the type definition to be incorrect for this field.

The metadata for this source should probably look something like
```json
{
   "bucket": "the bucket",
   "object": "whatever object is",
   "region": "US-EAST-1",
   "metadata": {
      "some_key_from_s3": "asdf",
      "region": "a key that was named region"
   }
}
```

Reply:
Would this just be updated to insert at `path!("metadata", key.as_str())`?